### PR TITLE
Update `LightningModule` docs

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

This is part of the docs revamp. We are looking for missing docs, broken links and embedded code sections in each component.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @justusschock @awaelchli @ananthsub @ninginthecloud @jjenniferdai

### CODE REVIEW

File: docs/source/common/lightning_module.rst [hunk 1]
```diff
@@ -267,10 +271,10 @@ The matching pseudocode is:
 
 Training with DataParallel
 ~~~~~~~~~~~~~~~~~~~~~~~~~~
-When training using an `accelerator` that splits data from each batch across GPUs, sometimes you might
+When training using an ``accelerator`` that splits data from each batch across GPUs, sometimes you might
```
Comment (kaushikb11) [inline docs/source/common/lightning_module.rst:274]:
```suggestion
When training using an ``strategy`` that splits data from each batch across GPUs, sometimes you might
```

File: docs/source/common/lightning_module.rst [hunk 2]
```diff
@@ -907,8 +952,8 @@ The current step (does not reset each epoch)
 
 hparams
 ~~~~~~~
-The arguments saved by calling ``save_hyperparameters`` passed through ``__init__()``
- could be accessed by the ``hparams`` attribute.
+The arguments saved by calling :meth:`~pytorch_lightning.core.mixins.HyperparametersMixin.save_hyperparameters`
```
Comment (kaushikb11) [inline docs/source/common/lightning_module.rst:955]:
```suggestion
The arguments saved by calling ``save_hyperparameters``
```
I see you did multiple changes to this format, I prefer the existing one. Imo, user doesn't need to know the reference

Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:955]: aren't references better? if they want to have a look at it.. then it's quick, no search required.

File: docs/source/common/lightning_module.rst [hunk 3]
```diff
@@ -139,6 +140,8 @@ The LightningModule has many convenience methods, but the core ones you need to
      - the full validation loop
    * - test_step
      - the full test loop
+   * - predict_step
```
Comment (kaushikb11) [inline docs/source/common/lightning_module.rst:143]: Thanks for adding the missing predict references.

File: docs/source/common/lightning_module.rst [hunk 4]
```diff
@@ -54,7 +55,7 @@ Notice a few things.
         new_x = torch.Tensor(2, 3)
         new_x = new_x.type_as(x)
 
-5.  Lightning by default handles the distributed sampler for you.
+5.  Lightning, by default, handles the distributed sampler for you when running under a distributed environment.
```
Comment (rohitgr7) [inline docs/source/common/lightning_module.rst:58]:
```suggestion
5.  Lightning, by default, handles the distributed sampler for you when running under a distributed strategy.
```

File: docs/source/common/lightning_module.rst [hunk 5]
```diff
@@ -54,7 +55,7 @@ Notice a few things.
         new_x = torch.Tensor(2, 3)
         new_x = new_x.type_as(x)
 
-5.  Lightning by default handles the distributed sampler for you.
+5.  Lightning, by default, handles the distributed sampler for you when running under a distributed strategy.
```
Comment (carmocca) [inline docs/source/common/lightning_module.rst:58]: I liked the "for you" statement more

**Review by carmocca — APPROVED** [review]

File: docs/source/common/lightning_module.rst [hunk 6]
```diff
@@ -379,10 +393,10 @@ If you need to do something with all the outputs of each `validation_step`, over
 
 Validating with DataParallel
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-When training using an `accelerator` that splits data from each batch across GPUs, sometimes you might
+When training using an ``strategy`` that splits data from each batch across GPUs, sometimes you might
```
Comment (ananthsub) [inline docs/source/common/lightning_module.rst:396]:
```suggestion
When training using a ``strategy`` that splits data from each batch across GPUs, sometimes you might
```

File: docs/source/common/lightning_module.rst (hunk 5 above)
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:58]: just trying to be a little more explicit about when actually lightning will handle distributed sampler.

File: docs/source/common/lightning_module.rst [hunk 7]
```diff
@@ -886,8 +931,8 @@ The device the module is on. Use it to keep your code device agnostic
 
 global_rank
 ~~~~~~~~~~~
-The global_rank of this LightningModule. Lightning saves logs, weights etc only from global_rank = 0. You
-normally do not need to use this property
+The global_rank of this LightningModule. Lightning saves logs, weights etc. only from global_rank = 0. You
+usually do not need to use this property
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:935]:
```suggestion
The ``global_rank`` is the index of the current process across all nodes and devices.
Lightning will perform some operations such as logging, weight checkpointing only when ``global_rank is 0``. You
usually do not need to use this property, but it is useful to know how to access it if needed.
```

File: docs/source/common/lightning_module.rst [hunk 8]
```diff
@@ -886,8 +931,8 @@ The device the module is on. Use it to keep your code device agnostic
...
 
 Global rank refers to the index of that GPU across ALL GPUs. For example, if using 10 machines, each with 4 GPUs,
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:937]:
```suggestion
For example, if using 10 machines, each with 4 GPUs,
```

File: docs/source/common/lightning_module.rst [hunk 9]
```diff
@@ -938,13 +983,12 @@ The current logger being used (tensorboard or other supported logger)
 
 local_rank
 ~~~~~~~~~~~
-The local_rank of this LightningModule. Lightning saves logs, weights etc only from global_rank = 0. You
-normally do not need to use this property
+The local_rank of this LightningModule. Lightning saves logs, weights etc. only from global_rank = 0. You
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:986]:
```suggestion
The ``local_rank`` of this LightningModule. Lightning saves logs, weights etc. only from global_rank = 0. You
```

Comment (tchaton) [reply docs/source/common/lightning_module.rst:986]: I will reformulate this. We are talking about `local_rank` and global_rank is used as above. This is confusing.

File: docs/source/common/lightning_module.rst [hunk 10]
```diff
@@ -1062,9 +1106,9 @@ Get the model file size (in megabytes) using ``self.model_size`` inside Lightnin
 truncated_bptt_steps
 ^^^^^^^^^^^^^^^^^^^^
 
-Truncated back prop breaks performs backprop every k steps of
+Truncated back prop breaks perform backprop every k steps of
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:1109]:
```suggestion
Truncated backpropagation through time performs backpropagation every k steps of
```

**Review by justusschock — APPROVED** [review]

File: docs/source/common/lightning_module.rst [hunk 11]
```diff
@@ -116,10 +117,10 @@ Which you can train by doing:
 .. code-block:: python
 
     train_loader = DataLoader(MNIST(os.getcwd(), download=True, transform=transforms.ToTensor()))
-    trainer = pl.Trainer()
+    trainer = pl.Trainer(max_epochs=1)
     model = LitModel()
 
-    trainer.fit(model, train_loader)
+    trainer.fit(model, train_dataloaders=train_loader)
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:123]: its `train_dataloaders`

File: docs/source/common/lightning_module.rst [hunk 12]
```diff
@@ -228,7 +231,8 @@ Here's the pseudocode of what it does under the hood:
 
 Train epoch-level operations
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-If you need to do something with all the outputs of each `training_step`, override `training_epoch_end` yourself.
+If you need to do something with all the outputs of each :meth:`~pytorch_lightning.core.lightning.LightningModule.training_step`.
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:234]: updated!

File: docs/source/common/lightning_module.rst [hunk 13]
```diff
@@ -267,10 +271,10 @@ The matching pseudocode is:
 
 Training with DataParallel
 ~~~~~~~~~~~~~~~~~~~~~~~~~~
-When training using an `accelerator` that splits data from each batch across GPUs, sometimes you might
+When training using a ``strategy`` that splits data from each batch across GPUs, sometimes you might
 need to aggregate them on the main GPU for processing (dp, or ddp2).
 
-In this case, implement the `training_step_end` method
+In this case, implement the :meth:`~pytorch_lightning.core.lightning.LightningModule.training_step_end`.
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:277]: updated!

File: docs/source/common/lightning_module.rst [hunk 14]
```diff
@@ -682,7 +696,7 @@ Then pass in any arbitrary model to be fit with this task
         task = ClassificationTask(model)
 
         trainer = Trainer(gpus=2)
-        trainer.fit(task, train_dataloader, val_dataloader)
+        trainer.fit(task, train_dataloaders=train_dataloader, val_dataloaders=val_dataloader)
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:699]: are you sure?
https://github.com/PyTorchLightning/pytorch-lightning/blob/ce0a977742872736f150f6d37ecaa301a318668f/pytorch_lightning/trainer/trainer.py#L693-L700

File: docs/source/common/lightning_module.rst (hunk 9 above)
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:986]: done!

File: docs/source/common/lightning_module.rst [hunk 15]
```diff
@@ -359,9 +363,19 @@ Under the hood, Lightning does the following:
...
+
+    model = Model()
+    trainer = Trainer()
+    trainer.validate(model)
+
 Validation epoch-level metrics
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-If you need to do something with all the outputs of each `validation_step`, override `validation_epoch_end`.
+If you need to do something with all the outputs of each :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_step`,
+override :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_epoch_end`
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:378]: done!

File: docs/source/common/lightning_module.rst [hunk 16]
```diff
@@ -782,6 +815,12 @@ save_hyperparameters
 .. automethod:: pytorch_lightning.core.lightning.LightningModule.save_hyperparameters
     :noindex:
 
+toggle_optimizer
```
[the remark this answers is not in the document]
Comment (carmocca) [reply docs/source/common/lightning_module.rst:820]: It's used in some code examples, i'd say yes.

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
Nice job fixing the missing references and methods!

File: docs/source/common/lightning_module.rst [hunk 17]
```diff
@@ -299,7 +303,7 @@ In this case, implement the `training_step_end` method
          for out in training_step_outputs:
              ...
 
-The full pseudocode that lighting does under the hood is:
+The complete pseudocode that lightning does under the hood is:
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst]: sure.. will take care for others too!
